File: backend/routers/tickers.py
"""
API routes for tickers.
"""
from fastapi import APIRouter, HTTPException, Query
from typing import Optional
from backend.models import Ticker, TickerCreate, TickerResponse, TickersListResponse
from backend.services import ticker_service

router = APIRouter(prefix="/api/tickers", tags=["tickers"])


@router.get("", response_model=TickersListResponse)
async def get_tickers():
    """
    Get all tickers.
    
    Returns all tickers sorted by symbol.
    """
    tickers = ticker_service.get_all_tickers()
    return TickersListResponse(
        success=True,
        tickers=tickers,
        count=len(tickers)
    )


@router.get("/{ticker_id}", response_model=TickerResponse)
async def get_ticker_by_id(ticker_id: int):
    """
    Get a ticker by its ID.
    
    - **ticker_id**: The ticker ID
    """
    ticker = ticker_service.get_ticker_by_id(ticker_id)
    if not ticker:
        raise HTTPException(status_code=404, detail=f"Ticker with id '{ticker_id}' not found")
    
    return TickerResponse(
        success=True,
        ticker=ticker,
        message="Ticker retrieved successfully"
    )


@router.get("/symbol/{symbol}", response_model=TickerResponse)
async def get_ticker_by_symbol(symbol: str):
    """
    Get a ticker by its symbol.
    
    - **symbol**: The ticker symbol (e.g., "AAPL")
    """
    ticker = ticker_service.get_ticker_by_symbol(symbol)
    if not ticker:
        raise HTTPException(status_code=404, detail=f"Ticker with symbol '{symbol}' not found")
    
    return TickerResponse(
        success=True,
        ticker=ticker,
        message="Ticker retrieved successfully"
    )


@router.post("", response_model=TickerResponse, status_code=201)
async def create_ticker(ticker_data: TickerCreate):
    """
    Create a new ticker.
    
    - **symbol**: The ticker symbol (e.g., "AAPL")
    - **name**: The company name
    - **sector**: The sector
    """
    try:
        # Check if ticker already exists
        existing = ticker_service.get_ticker_by_symbol(ticker_data.symbol)
        if existing:
            raise HTTPException(status_code=400, detail=f"Ticker with symbol '{ticker_data.symbol}' already exists")
        
        ticker = ticker_service.create_ticker(
            symbol=ticker_data.symbol,
            name=ticker_data.name,
            sector=ticker_data.sector
        )
        
        if not ticker:
            raise HTTPException(status_code=500, detail="Failed to create ticker")
        
        return TickerResponse(
            success=True,
            ticker=ticker,
            message="Ticker created successfully"
        )
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error creating ticker: {str(e)}")

File: frontend/src/api/tickers.js
/**
 * API service for tickers
 * Connects to the FastAPI backend
 */

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || "http://localhost:8000";

/**
 * Get all tickers
 * @returns {Promise<{success: boolean, tickers?: Array, count?: number, error?: string}>}
 */
export async function getTickers() {
  try {
    const response = await fetch(`${API_BASE_URL}/api/tickers`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });
    
    if (!response.ok) {
      throw new Error(`Failed to fetch tickers: ${response.status} ${response.statusText}`);
    }
    
    const data = await response.json();
    return { success: true, tickers: data.tickers || [], count: data.count || 0 };
  } catch (error) {
    console.error("Error fetching tickers:", error);
    return { success: false, error: error.message, tickers: [], count: 0 };
  }
}

/**
 * Get a ticker by ID
 * @param {number} tickerId - The ticker ID
 * @returns {Promise<{success: boolean, ticker?: Object, error?: string}>}
 */
export async function getTickerById(tickerId) {
  try {
    const response = await fetch(`${API_BASE_URL}/api/tickers/${tickerId}`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });
    
    if (!response.ok) {
      if (response.status === 404) {
        return { success: false, error: "Ticker not found", ticker: null };
      }
      throw new Error(`Failed to fetch ticker: ${response.status} ${response.statusText}`);
    }
    
    const data = await response.json();
    return { success: true, ticker: data.ticker };
  } catch (error) {
    console.error("Error fetching ticker:", error);
    return { success: false, error: error.message, ticker: null };
  }
}

/**
 * Get a ticker by symbol
 * @param {string} symbol - The ticker symbol (e.g., "AAPL")
 * @returns {Promise<{success: boolean, ticker?: Object, error?: string}>}
 */
export async function getTickerBySymbol(symbol) {
  try {
    const response = await fetch(`${API_BASE_URL}/api/tickers/symbol/${encodeURIComponent(symbol)}`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });
    
    if (!response.ok) {
      if (response.status === 404) {
        return { success: false, error: "Ticker not found", ticker: null };
      }
      throw new Error(`Failed to fetch ticker: ${response.status} ${response.statusText}`);
    }
    
    const data = await response.json();
    return { success: true, ticker: data.ticker };
  } catch (error) {
    console.error("Error fetching ticker by symbol:", error);
    return { success: false, error: error.message, ticker: null };
  }
}

/**
 * Create a new ticker
 * @param {Object} tickerData - {symbol, name, sector}
 * @returns {Promise<{success: boolean, ticker?: Object, error?: string}>}
 */
export async function createTicker(tickerData) {
  try {
    const response = await fetch(`${API_BASE_URL}/api/tickers`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(tickerData),
    });
    
    if (!response.ok) {
      throw new Error(`Failed to create ticker: ${response.status} ${response.statusText}`);
    }
    
    const data = await response.json();
    return { success: true, ticker: data.ticker };
  } catch (error) {
    console.error("Error creating ticker:", error);
    return { success: false, error: error.message, ticker: null };
  }
}

